src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/ForkScanner.java

```diff
@@ -407,27 +382,30 @@ ArrayDeque<ParallelBlockStart> leastCommonAncestor(@NonNull final Set<FlowNode>
                 FlowPiece existingPiece = branches.get(nextBlockStart);
                 if (existingPiece == null && myPiece instanceof FlowSegment) { // No merge, just add to segment
                     ((FlowSegment) myPiece).add(nextBlockStart);
-                    branches.put(nextBlockStart, myPiece);
+                    if (isParallelStart(nextBlockStart)) {
+                        // We make a fork for every parallel step preemptively in case it has less than 2 branches.
```

Comment:
This is the fix, essentially instead of waiting until we have a piece that needs to be merge with an existing segment, we create the `Fork` objects right away as soon as we see a `parallel` step.